### PR TITLE
calculate_average

### LINKED ISSUE #50: calculate_average
Create a function that takes a dictionary of student marks and a student's name as input, calculates the average of their marks, handles potential errors if the student is not found, and returns the rounded average marks

### PR DESCRIPTION
**Pull request by mariacommits**
--- #50 
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_calculate_average.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+"""This a unit test for calculate_average function"""
+
+import unittest
+
+from ..calculate_average import calculate_average
+
+
+class TestGetAverage(unittest.TestCase):
```
Comment (ermishina) [inline solutions/tests/test_calculate_average.py:8]: Add the test class has a docstring

File: solutions/calculate_average.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+def calculate_average(student_marks: dict, query_name: str) -> float:
+    """Calculates the average marks for a given student.
+
+    Args:
+        student_marks: A dictionary where keys are student name
+
+        query_name: The name of the student for whom to calculate the average.
+
+    Returns:
+        The average marks of the student, rounded to 2 decimal places.
```
Comment (ermishina) [inline solutions/calculate_average.py:24]: The type of the return is a guess through the rounded of 2 decimal. I would ensure that the Returns section explicitly mentions the type float in the text (even though it is in the type annotation).

File: solutions/calculate_average.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    Raises:
+        KeyError: If the query_name is not found in the student_marks dictionary.
```
Comment (ermishina) [inline solutions/calculate_average.py:27]: KeyError is raised for missing student names, which is appropriate. Consider adding a specific message to make debugging easier. Example:
raise KeyError(f"Student '{query_name}' not found in records.")

File: solutions/calculate_average.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        26.5
+
+        >>> student_marks = {'Alice': [90, 85, 92], 'Bob': [78, 82, 75]}
+        >>> calculate_average(student_marks, 'Alice')
+        89.0
+
+        >>> student_marks = {'David': [65, 70, 68], 'Emily': [88, 91, 85]}
+        >>> calculate_average(student_marks, 'Emily')
+        88.0
+    """
```
Comment (ermishina) [inline solutions/calculate_average.py:45]: The examples in the docstring are clear and provide good context for understanding the function's usage. The spacing between sections in the function enhances readability.

File: solutions/tests/test_calculate_average.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        """
+        student_records = {"Charlie": [87.5, 92.25, 89.75]}
+        self.assertEqual(calculate_average(student_records, "Charlie"), 89.83)
+
+    def test_get_average_large_number_of_marks(self):
+        """
+        Tests if the average is correctly calculated for a student with a large number of marks.
+        """
+        student_records = {"David": [90] * 100}  # 100 marks of 90
+        self.assertEqual(calculate_average(student_records, "David"), 90.00)
```
Comment (ermishina) [inline solutions/tests/test_calculate_average.py:66]: The test file has good coverage of scenarios, including edge cases like negative marks, empty dictionaries, and large data sets. However you can Include test cases for invalid input types (e.g., a student name as an integer or marks as a string).

**Review by ermishina — CHANGES_REQUESTED** [review]
The code is well-structured, easy to read, and follows basic formatting principles.
The function documentation is clear, includes a description, arguments, and return values, making it helpful for users.
The test cases cover a wide range of scenarios, including edge cases (e.g., empty records or negative marks), which improves the reliability of the function.
The use of type annotations makes the code more understandable and reduces the chance of errors.
Recommendations:
Improve error messages: Add more detailed messages for KeyError to make debugging easier.
Cover incorrect input types with tests: Add tests to check cases where input data has the wrong type (e.g., the student name is a number).
Specify the float type in the Returns section of the documentation: Even though the type annotation is present, it’s helpful to explicitly mention this in the description for clarity.

**Review by ermishina — APPROVED** [review]
